### PR TITLE
Goto support

### PR DESCRIPTION
**Pull request by derek-pryor**
Adding support for a new `Goto("state name")` state that will allow jumping to any state within the branch of execution. Added a check to verify that the named state is valid.

Added for mainly for the use case of common error handling
```
Lambda("function1")
    catch []: '$.error'
        Goto("handler")
        
Lambda("function2")
    catch []: '$.error'
        Goto("handler")
        
Lambda("function3")
    catch []: '$.error'
        Goto("handler")
        
Success()

Lambda("handler")
    """handler
    common error handler function
    """
    
Fail('error', 'cause')
```
----
Edit: Changed the syntax from `Goto('state name')` to `goto 'state name'`

### CODE REVIEW

**Comment by sandyhider** [comment]
> Is it going to workout Ok if the script does a goto into the middle of an ASTStateParallel state?

Ok I see that error_invalid_goto_target.sfn actually doesn't allow parallel state being used in a jump.  Thats good.

File: docs/StepFunctionDSL.md [hunk 1]
```diff
@@ -171,6 +172,15 @@ Modifiers:
             raw input (before the `input` modifier was applied) (Default: `"$"`)
 * `data`: A block of Json data that will be used as the result of the state
 
+#### Goto State
+A state that does nothing `Goto()` can be used to transfer control to the named
+state. This is can be to create common error handling routines (among other
```
Comment (sandyhider) [inline docs/StepFunctionDSL.md]: This is can

**Comment by derek-pryor** [comment]
> 
> 
> Is Goto A State or a Flow Control? ... It looks like it is clearly a state. Perhaps add something in the documentation to explain why this is a state and not a Flow Control?

Yea, I realized that after I implemented it. I was thinking about changing it to `goto "state name"` to be more like a flow control structure...

**Comment by derek-pryor** [comment]
> 
> 
> Is it going to workout Ok if the script does a goto into the middle of an ASTStateParallel state?

You cannot do that, as a parallel branches states are self contained. You also cannot goto from a parallel branch to outside of the parallel statement or another branch of the parallel statement. The target verification check helps prevent this.

File: heaviside/ast.py [hunk 2]
```diff
@@ -507,3 +541,36 @@ def resolve_arns(branch, translate = lambda x, y: y):
...
+        branch.raise_error("Trying to check goto targets for non-branch state")
+
+    to_process = [branch.states]
+
+    while len(to_process) > 0:
+        states = to_process.pop(0)
+
+        names = set() # Need to know all of the valid state names for the branch
+        for state in states:
+            names.add(state.name)
```
Comment (sandyhider) [inline heaviside/ast.py]: Are we checking somewhere else that this branch of states doesn't contain two states with the same name?  I don't see that check in this function?

Comment (derek-pryor) [reply heaviside/ast.py]: Line 481 `def check_names(branch)` is the check that verifies state names. It is run before this check, but could be run afterwards. I wanted to keep the check functions simple and didn't want to co-mingle the logic of the two checks.

Comment (sandyhider) [reply heaviside/ast.py]: Cool.

**Pull request merged**